tests/tests/test_mender_shell.py

```diff
@@ -0,0 +1,91 @@
+# Copyright 2020 Northern.tech AS
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        https://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+import logging
+import pytest
+
+from testutils.api import proto_shell
+from testutils.api.proto_shell import WRONG_BEHAVIOR
+from ..common_setup import standard_setup_one_client_bootstrapped
+from ..MenderAPI import devconnect
+
+if WRONG_BEHAVIOR:
+    logger = logging.getLogger("websockets")
+    logger.setLevel(logging.DEBUG)
+    logger.addHandler(logging.StreamHandler())
+
+
+class TestMenderShell:
+    def test_regular_protocol_commands(self, standard_setup_one_client_bootstrapped):
+        device = standard_setup_one_client_bootstrapped.device
+
+        if WRONG_BEHAVIOR:
+            # MEN-4273.
+            device.run(
+                """sed -i -e '2i"SkipVerify": true,' /etc/mender/mender-shell.conf"""
+            )
+            device.run("systemctl restart mender-shell")
+
+        with devconnect.get_websocket() as ws:
+            # Start shell.
+            shell = proto_shell.ProtoShell(ws)
+            shell.startShell()
+
+            # Drain any initial output from the prompt. It should end in either "# "
+            # (root) or "$ " (user).
+            output = shell.recvOutput()
+            assert output[-2:].decode() in [
+                "# ",
+                "$ ",
+            ], "Could not detect shell prompt."
+
+            if not WRONG_BEHAVIOR:
+                # Starting the shell again should be a no-op. It should return that it
+                # is already started, as long as the shell limit is 1.
+                # MEN-4240.
+                shell.startShell()
+
+            # Test if a simple command works.
+            shell.sendInput("ls /\n".encode())
+            output = shell.recvOutput()
+            output = output.decode()
+            assert "usr" in output
+            assert "etc" in output
+
+            # Try to stop shell.
+            shell.stopShell()
+
+            # Make sure we can not send anything to the shell.
+            shell.sendInput("ls /\n".encode())
+            output = shell.recvOutput()
+            output = output.decode()
+            assert "usr" not in output
+            assert "etc" not in output
```

Comment:
Won't we get an error, somehow?

Reply:
Good point, I think this deserves some discussion. I started the discussion in [MEN-4241](https://tracker.mender.io/browse/MEN-4241?focusedCommentId=111749&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-111749) so that it's easier to track the discussion.

For the time being, I will not take this into account in my next update.